### PR TITLE
stable trace state

### PR DESCRIPTION
**Pull request by sorenoid**
Related to issue: https://devtopia.esri.com/runtime/kotlin/issues/4652

### Description:

annotate TraceState as stable
make TraceStaet public/internal property types State<T>
push navController into the TraceNavigation composable because it is an unstable parameter
adjust some navigation logic that used a StateFlow (now uses a State object)

### Pre-merge Checklist

- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/view/vTest/job/vtest/job/toolkit/108/
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [ ] Yes
  - [ ] No

### CODE REVIEW

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/AddStartingPointScreen.kt [hunk 1]
```diff
@@ -36,19 +34,16 @@ import com.arcgismaps.toolkit.utilitynetworks.TraceState
...
 @Composable
 internal fun AddStartingPointScreen(traceState: TraceState, onStopPointSelection: () -> Unit) {
-    LaunchedEffect(key1 = traceState.addStartingPointMode) {
-        traceState.addStartingPointMode.collect {
-            if (it == AddStartingPointMode.Stopped) {
-                onStopPointSelection()
-            }
-        }
-    }
+    println("UTILITY NETWORK AddStartingPointScreen")
```
Comment (puneet-pdx) [inline toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/AddStartingPointScreen.kt:37]: this could be removed

**Review by puneet-pdx — COMMENTED** [review]
@sorenoid with these changes I am seeing a performance issue, where on initial load the trace tool has obvious blink effect. Where I think the trace tool is entering the composition and then quickly recomposing.

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt [hunk 2]
```diff
@@ -393,6 +408,15 @@ internal data class StartingPoint(val feature: ArcGISFeature, val utilityElement
         symbol.createSwatch(screenScale).getOrThrow()
 }
 
+@Immutable
+internal data class TraceRun(
+    val name: String, // need to auto populate this, if not provided by AdvancedOptions
+    val graphics: List<Graphic>,
+    val featureResults: List<UtilityElement>,
+    val functionResults: List<UtilityTraceFunctionOutput>,
+    val geometryResults: List<Graphic>
```
Comment (puneet-pdx) [inline toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt:417]: `geometryResults` are not needed as they are the same as graphics. I had removed these in my last PR

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt [hunk 3]
```diff
@@ -385,6 +388,18 @@ public sealed class AddStartingPointMode {
     public data object None : AddStartingPointMode()
 }
 
+/**
+ * Defines a navigation route for the trace tool screens.
+ *
+ * @since 200.6.0
+ */
+internal enum class TraceNavRoute {
```
Comment (puneet-pdx) [inline toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt:396]: What is the advantage of having this in `TraceState`? We should just have it in `TraceNaviagation`

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/TraceResultScreen.kt [hunk 4]
```diff
@@ -64,7 +64,8 @@ internal fun TraceResultScreen(
     traceRun: TraceRun,
     onDeleteResult: () -> Unit,
     onZoomToResults: () -> Unit,
-    onClearAllResults: () -> Unit
+    onClearAllResults: () -> Unit,
+    onDismiss: () -> Unit
```
Comment (puneet-pdx) [inline toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/TraceResultScreen.kt:68]: `onDismiss` is not used, could be removed ?

**Review by puneet-pdx — COMMENTED** [review]
@sorenoid few questions/comments

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/Trace.kt [hunk 5]
```diff
@@ -42,39 +42,42 @@ internal const val traceSurfaceContentDescription: String = "trace component sur
...
     modifier: Modifier = Modifier
 ) {
-    LaunchedEffect(key1 = traceState) {
-        traceState.initialize()
-    }
+    val initializationStatus by traceState.initializationStatus
 
-    val initializationStatus = traceState.initializationStatus.collectAsStateWithLifecycle()
+    LaunchedEffect(traceState) {
+        if (initializationStatus is InitializationStatus.NotInitialized
```
Comment (puneet-pdx) [inline toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/Trace.kt:55]: Do we need this check here ? We are already handling it in [TraceState](https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/utilitynetworktrace/toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt#L121)

File: toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt (hunk 3 above)
Comment (sorenoid) [reply toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt:396]: Previously we had two sources of truth governing navigation, we had the add starting point mode, and we had this enum. This was leading to complex logic in `TraceNavigation`. Now we just have one source of truth governing navigation. With one source of truth, we need to affect it outside of the composition sometimes, so the nav state is persisted in the state object. For example we need to change the nav route from `TraceState.identifyFeatures`.

File: toolkit/utilitynetworks/build.gradle.kts [hunk 6]
```diff
@@ -82,6 +82,11 @@ android {
     }
 }
 
+composeCompiler {
```
[the remark this answers is not in the document]
Comment (sorenoid) [reply toolkit/utilitynetworks/build.gradle.kts:85]: I'll remove it, and can add it back when necessary.

**Comment by sorenoid** [comment]
@puneet-pdx I've addressed your comments, making some code changes and continuing conversations in the PR. Can you please re-review or continue the discussion threads?

**Review by puneet-pdx — APPROVED** [review]
@sorenoid looks good 👍🏽

**Pull request merged**